GVFS/GVFS.Common/Git/GitSsl.cs

```diff
@@ -0,0 +1,208 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text.RegularExpressions;
+using GVFS.Common.Tracing;
+
+namespace GVFS.Common.Git
+{
+    public class GitSsl
+    {
+        private readonly string certificatePathOrSubjectCommonName;
+        private readonly bool isCertificatePasswordProtected;
+
+        public GitSsl()
+        {
+            this.certificatePathOrSubjectCommonName = null;
+            this.isCertificatePasswordProtected = false;
+            this.ShouldVerify = true;
+        }
+
+        public GitSsl(IDictionary<string, GitConfigSetting> configSettings) : this()
+        {
+            if (configSettings != null)
+            {
+                if (configSettings.TryGetValue(GitConfigSetting.HttpSslCert, out GitConfigSetting sslCerts))
+                {
+                    this.certificatePathOrSubjectCommonName = sslCerts.Values.Single();
+                }
+
+                if (configSettings.TryGetValue(GitConfigSetting.HttpSslCertPasswordProtected, out GitConfigSetting isSslCertPasswordProtected))
+                {
+                    this.isCertificatePasswordProtected = isSslCertPasswordProtected.Values.Select(bool.Parse).Single();

```

Comment:
Similar question for `bool.Parse`, what will happen if it throws?

Reply:
If we have this git config
```
git config --get-urlmatch http https://localhost:8081 

http.sslcert testCert2 test
http.sslverify fake news
```

Then when you invoke git, this is what you get
```
git ls-remote https://localhost:8081
fatal: bad numeric config value 'fake news' for 'http.sslverify': invalid unit
```

So we should fail, but the question is, if we should catch, and log a friendlier message.

Reply:
> the question is, if we should catch, and log a friendlier message

Correct me if I'm wrong but it looks like:

- `GitSsl` is only created in the `GitAuthentication` constructor
- The `GitAuthentication` constructor is only called from the `Enlistment` constructor.

I've taken a look at some of the paths that hit the `Enlistment` constructor, and it looks like they *only* expect it to throw `InvalidRepoException` (and they expect the exception to contain a helpful error message).

*Example from `Enlistment` constructor*

```
GitProcess.ConfigResult originResult = gitProcess.GetOriginUrl();
if (!originResult.TryParseAsString(out string originUrl, out string error))
{
    throw new InvalidRepoException("Could not get origin url. git error: " + error);
}

if (originUrl == null)
{
    throw new InvalidRepoException("Could not get origin url. remote 'origin' is not configured for this repo.'");
}
```

I think we should maintain that contract (only throwing helpful `InvalidRepoException` messages from the `Enlistment` constructor), and so I think we have a few options:

- Move the code that can throw out of the constructor and into a `TryInitialize` type method
- Update `GitSsl` to catch the exception that can get thrown in its constructor and throw `InvalidRepoException` exceptions instead (with a helpful error message)

We could also catch everything in the `Enlistment` constructor and convert it to a `InvalidRepoException`, but my concern is that message will not have as much context for the user (and these are user facing messages):

*Example of code that handles `InvalidRepoException`*

```
catch (InvalidRepoException e)
{
    this.ReportErrorAndExit(
        "Error: '{0}' is not a valid GVFS enlistment. {1}",
        enlistmentRootPath,
        e.Message);
}
```

Reply:
Makes sense, will make sure we throw exceptions with messages, that are helpful to the user.

Reply:
Looked at uses - it seems CloneVerb only handles generic exceptions, not `InvalidRepoException`:
https://github.com/Microsoft/VFSForGit/blob/master/GVFS/GVFS/CommandLine/CloneVerb.cs#L244-L261

Added handling for this exception and started throwing it from GitSsl constructor, when bool parsing fails.